### PR TITLE
Fix Evolution: Current Depos. and Shapes

### PR DESCRIPTION
**Pull request by mrowan137**
This PR (the first of 2; see PR #1084 for the continuation) introduces changes to give better agreement between single precision (SP) and double precision (DP) runs.  In particular, this PR introduces changes in the current deposition; these changes affect the early time evolution, and improve the 'missing currents' in single precision (this is demonstrated in the plot below).  Still, these changes have little effect on the late-time evolution (see the attached videos, also the plot from PR #1084).

[For continuity with previous discussions on the issue, the plot below (also the videos further below) use the convention jx-->jy]
dp_sp_compare_jy

Functions to compute shape factors are recast as functors (to avoid would-be code duplication), which allows to compute shape factors in SP or DP (depending on the arguments' type); DP shape factors and positions do not appear to be necessary in FieldGather to ensure good agreement between SP and DP (so when the code is compiled in SP mode, the shape factor functors get DP arguments in CurrentDeposition, whereas the safeguard is not necessary in FieldGather).

Single precision benchmarks are reset where necessary (these change because the SP results now use DP in CurrentDeposition, whereas the benchmarks previously used SP).

Videos are attached to show the agreement between SP and DP with and without both fixes (including also those from PR #1084), for both Esirkepov and direct deposition:
[movies.zip](https://github.com/ECP-WarpX/WarpX/files/4756095/movies.zip)

Input file to reproduce runs from the videos is here: 
[inputs.txt](https://github.com/ECP-WarpX/WarpX/files/4935321/inputs.txt)

### CODE REVIEW

File: Source/Particles/ShapeFactors.H [hunk 1]
```diff
@@ -19,6 +25,13 @@ int compute_shape_factor(amrex::Real* const sx, amrex::Real xint)
     return 0;
 };
 
+template <int depos_order>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shape_factor_dbl(double* const sx, double xint)
```
Comment (ax3l) [inline Source/Particles/ShapeFactors.H]: You could probably introduce those as templates, as they are inline header functions already :)
```C++
template <int depos_order, typename T=amrex::Real>
AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
int compute_shape_factor(T* const sx, T xint)
// ...
```
called as:
```C++
compute_shape_factor<depos_order>( /* ... */ ) // Real
compute_shape_factor<depos_order, double>( /* ... */ ) // double
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 2]
```diff
@@ -135,67 +135,87 @@ void doDepositionShapeN(const GetParticlePosition& GetPosition,
             // x direction
             // Get particle position after 1/2 push back in position
 #if (defined WARPX_DIM_RZ)
-            const amrex::Real xmid = (rpmid - xmin)*dxi;
+            const double xmid = (rpmid - xmin)*dxi;
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H:139]: please don't forget to add comments before explicitly selected precision lines :)

**Review by ax3l — COMMENTED** [review]
Although the PR is WIP, I added a few comments to avoid code duplication.

Adding the tests that you did to CI would be wonderful so we can refactor the mixed-precision code to use more single-precision later on.

File: Source/Particles/Deposition/CurrentDeposition.H (hunk 2 above)
Comment (mrowan137) [reply Source/Particles/Deposition/CurrentDeposition.H:139]: Comments added!

File: Source/Particles/ShapeFactors.H [hunk 3]
```diff
@@ -11,166 +11,183 @@
  *  Compute shape factor and return index of leftmost cell where
  *  particle writes.
  *  Specialized templates are defined below for orders 0 to 3.
+ *  The `double` version of shape factor functions can be used
+ *  in current deposition to ensure that current deposited by
+ *  particles that move only a small distance is still resolved.
+ *  Without this, single and double precision version of the code
+ *  can give disagreeing results in the time evolution for some
+ *  problem setups.
```
Comment (mrowan137) [inline Source/Particles/ShapeFactors.H]: @mrowan137 fix this comment

File: Source/Particles/ShapeFactors.H (hunk 1 above)
Comment (mrowan137) [reply Source/Particles/ShapeFactors.H]: These are now introduced as functors, like you suggested and outlined, thanks again for explaining the details!

**Comment by ax3l** [comment]
@mrowan137 is it possible to check the walltime (without init) for the `inputs_full_single_WarpX_20-07 ` file with this PR on Summit's V100?
https://github.com/ax3l/picongpu/tree/topic-20200714-thermalBench/share/picongpu/examples/ThermalBenchmark

@atmyers and I just took the benchmark two days ago and I wonder how much we loose in SP, if any at all.

File: Source/Particles/ShapeFactors.H [hunk 4]
```diff
@@ -11,166 +11,183 @@
...
-    return j;
-}
+struct Compute_shape_factor< 0 >
+{
+    template< typename T >
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    int operator()(T* const sx, T xmid) const
+    {
+        const auto j = static_cast<int>(xmid + (T)0.5);
+        sx[0] = (T)1.0;
```
Comment (ax3l) [inline Source/Particles/ShapeFactors.H]: Exemplified here:
```suggestion
        sx[0] = T(1.0);
```

This is what differentiates us from `C` :sweat_smile: 
(Please don't say `static_cast` now.)

Comment (mrowan137) [reply Source/Particles/ShapeFactors.H]: Done :-)

**Comment by mrowan137** [comment]
@ax3l @atmyers 
> @mrowan137 is it possible to check the walltime (without init) for the `inputs_full_single_WarpX_20-07 ` file with this PR on Summit's V100?
> https://github.com/ax3l/picongpu/tree/topic-20200714-thermalBench/share/picongpu/examples/ThermalBenchmark
> 
> @atmyers and I just took the benchmark two days ago and I wonder how much we loose in SP, if any at all.

I did some tests (8 particles per cell to increase the memory usage), the measured times are the Evolve loop inclusive times (not clear how robust is this time difference between the mixed DP-SP and full SP)

```
             | time (s)     | used memory (GB)
Full DP      | 65.48        | 13
Mixed DP-SP  | 58.3         | 8
Full SP      | 53.3         | 8
```

**Comment by ax3l** [comment]
This looks pretty solid, thank you for adding those! Just for completeness, can you add how much time field gather and current deposition took relative to the evolve loop in full SP (I can look it up as well otherwise)?

**Review by ax3l — APPROVED** [review]
This is great work, thank you a lot! :rocket: :sparkles:

**Pull request merged**